### ISSUE TITLE
FunctionTimedOut exception cannot be raised with zero arguments

### ISSUE THREAD

**Issue by GrimzEcho** [kata198/func_timeout#6]
The function signature for `func_timeout.exceptions.FunctionTimedOut` is 

```python
def __init__(self, msg='', timedOutAfter=None, timedOutFunction=None, timedOutArgs=None, timedOutKwargs=None):
```
Indicating that all parameters are optional.

However, creating or throwing the exception with zero parameters raises an `AttributeError` exception

Minimal example

```python

def throw_timeout_with_message():
    raise FunctionTimedOut('foo')

def throw_blank_timeout():
    raise FunctionTimedOut()

if __name__ == '__main__':
    # works correctly
    try:
        throw_timeout_with_message()
    except FunctionTimedOut as e:
        assert(isinstance(e, FunctionTimedOut))
        print('handled')

    # raises an uncaught error
    try:
        throw_blank_timeout()
    except FunctionTimedOut as e:
        assert(isinstance(e, FunctionTimedOut))
        print('handled')

```

#### Expected result of the second try/catch block:
`> handled`

#### Actual result of the second try/catch block:
```
Traceback (most recent call last):
  File ".../scratch_3.py", line 40, in <module>
    throw_blank_timeout()
  File ".../scratch_3.py", line 8, in throw_blank_timeout
    raise FunctionTimedOut()
  File "...\Python27\lib\site-packages\func_timeout\exceptions.py", line 36, in __init__
    msg = self.getMsg()
  File "...\Python27\lib\site-packages\func_timeout\exceptions.py", line 49, in getMsg
    return 'Function %s (args=%s) (kwargs=%s) timed out after %f seconds.\n' %(self.timedOutFunction.__name__, repr(self.timedOutArgs), repr(self.timedOutKwargs), self.timedOutAfter)
AttributeError: 'NoneType' object has no attribute '__name__'
```

**Comment by kata198** [comment]
So... what exactly is your intent here? This is operating as designed, the exception is not intended to be raised directly by you.

The default behaviour is that it is raised from within the context of a timed out function -- that is the real issue, there is no function attached because you are not using it through the public API. So it cannot generate the timeout message.

The fix is simple, can just put a try/except around line 36 of exceptions.py, like so:

    --- a/func_timeout/exceptions.py
    +++ b/func_timeout/exceptions.py
    @@ -33,7 +33,10 @@ class FunctionTimedOut(BaseException):
             self.timedOutKwargs = timedOutKwargs
 
             if not msg:
    -            msg = self.getMsg()
    +            try:
    +                msg = self.getMsg()
    +            except:
    +                msg = ''
     
             BaseException.__init__(self, msg)
 
But to do so I need to understand what you are trying to accomplish. Why are you trying to raise the exception directly, instead of using the public API?

**Comment by GrimzEcho** [comment]
Mainly for mocking and testing code that can throw the TimedOut exception. In our project, we need to add a timeout to a command invocation function controlled by a third-party library. To test that our code responds correctly to a timeout, we mock the command invocation and have it raise FunctionTimedOut in response to certain scenarios.

It's not a big issue, but I would argue that the exception is public and that its signature is somewhat misleading since it some of the parameters marked optional really aren't.

**Comment by kata198** [comment]
Okay, I understand. Sorry, I'm not getting alerted to these replies for some reason...

Anyway, yes, the exception itself is public, because it is catchable. If you're mocking it for testing, is there a reason you can't provide either your own message, or link it to a function?

The intended interface is that either a message is provided, or a function is provided (to generate a message). As python doesn't have a way to specify "either this argument, or this other argument" in the signature itself, the only other way I could enforce such is by raising an exception with a specific error message if one or the other is not satisfied. But that doesn't get you any further.

How about this... I just release 4.3.4 to squash a deprecation warning, I'll release 4.3.5 directly after and move the "self.timedOutFunction.__name__" from being inline in the message string generation, to being explicitly checked and assign the relevant string to "unknown function" if it is not set. That should work for your use case, yes? Watch for 4.3.5 in a couple minutes, and it should allow a default constructor to work.

**Comment by GrimzEcho** [comment]
Thanks for the response. You are right, it is not difficult to supply a string when instantiating the TimedOut exception. And I see what you mean about specifying one or the other. 

I actually forked your code for our project, as we needed the behavior of allowOverride/forceTimeout to be the default (e.g. the wrapper expected that the wrapped function will have a forceTimeout value to control the timeout interval, and won't modify the function or start a thread if the argument is not passed).

In the process, I also implemented the change you just suggested. Modifying TimeoutExecption to generate a generic string if neither the string nor the function was provided.

Thanks!  We tried two other libraries before finding one (yours) that would reliably interrupt a running process (that we have no control over) on both Windows and Linux.

**Comment by kata198** [comment]
I just released 4.3.5, which will allow the pattern you are requiring. The following unit test now passes:

    def test_instantiateExceptionNoArgs(self):

        gotException = False

        try:
            exc = FunctionTimedOut()
            msg = str(exc)
            msg2 = exc.getMsg()

        except Exception as _e:
            sys.stderr.write('Got unexpected exception in test_instantiateExceptionNoArgs with no arguments. %s  %s\n\n' %(str(type(_e)), str(_e)))
            gotException = True

        assert gotException is False, 'Expected to be able to create FunctionTimedOut exception without arguments.'

If no message is provided, it will try to generate a message. If no function is attached, it will replace the function name with "Unknown Function" and if the timedOutAfter is also missing, it will replace the float seconds of timeout with a string, "Unknown".

Feel free to merge in these changes into your fork!

Btw, if you'd like to just extend rather than fork, you can get all the updates available.

Here's how you would extend to change the default:

    from func_timeout import func_set_timeout as orig_func_set_timeout

    def func_set_timeout(timeout, allowOverride=True):
        return orig_func_set_timeout( timeout=timeout, allowOverride=allowOverride )

But do it however you feel is best!

Let me know if this resolves your issue, and I'll close the ticket. If you're able to close it yourself, you may do so also.

Thanks!

**Comment by GrimzEcho** [comment]
Thank you. This resolves the issue.
